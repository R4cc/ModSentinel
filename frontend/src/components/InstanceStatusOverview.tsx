import React from "react";
import { CheckCircle, RefreshCcw, XCircle } from "lucide-react";
import { cn } from "@/lib/utils.js";

type Labels = {
  upToDate?: string;
  updatesAvailable?: string;
  failed?: string;
};

export type InstanceStatusOverviewProps = {
  upToDate: number;
  updatesAvailable: number;
  failed: number;
  labels?: Labels;
  className?: string;
};

function StatCard({
  icon,
  value,
  label,
  color,
}: {
  icon: React.ReactNode;
  value: number;
  label: string;
  color: "emerald" | "amber" | "red";
}) {
  const colorClasses = {
    emerald: {
      icon: "text-emerald-600",
      value: "text-emerald-700",
    },
    amber: {
      icon: "text-amber-600",
      value: "text-amber-700",
    },
    red: {
      icon: "text-red-600",
      value: "text-red-700",
    },
  } as const;

  const c = colorClasses[color];

  return (
    <div className="rounded-md border p-sm shadow-sm">
      <div className="flex items-center gap-sm">
        <span className={cn("shrink-0", c.icon)} aria-hidden>
          {icon}
        </span>
        <div className="flex items-baseline gap-xs">
          <span className={cn("text-3xl font-bold leading-none", c.value)}>
            {value}
          </span>
          <span className="text-sm text-muted-foreground">{label}</span>
        </div>
      </div>
    </div>
  );
}

export default function InstanceStatusOverview({
  upToDate,
  updatesAvailable,
  failed,
  labels,
  className,
}: InstanceStatusOverviewProps) {
  const l: Required<Labels> = {
    upToDate: labels?.upToDate ?? "Up to date",
    updatesAvailable: labels?.updatesAvailable ?? "Updates available",
    failed: labels?.failed ?? "Failed",
  } as Required<Labels>;

  return (
    <div
      className={cn(
        "grid grid-cols-1 gap-sm sm:grid-cols-3",
        className,
      )}
      role="region"
      aria-label="Instance status overview"
    >
      <StatCard
        icon={<CheckCircle className="h-6 w-6" />}
        value={upToDate}
        label={l.upToDate}
        color="emerald"
      />
      <StatCard
        icon={<RefreshCcw className="h-6 w-6" />}
        value={updatesAvailable}
        label={l.updatesAvailable}
        color="amber"
      />
      <StatCard
        icon={<XCircle className="h-6 w-6" />}
        value={failed}
        label={l.failed}
        color="red"
      />
    </div>
  );
}

