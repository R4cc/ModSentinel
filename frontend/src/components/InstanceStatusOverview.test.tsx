import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { describe, it, expect } from "vitest";
import InstanceStatusOverview from "@/components/InstanceStatusOverview.tsx";

describe("InstanceStatusOverview", () => {
  it("renders numbers and labels correctly", () => {
    render(
      <InstanceStatusOverview upToDate={10} updatesAvailable={3} failed={1} />,
    );

    // Numbers
    expect(screen.getByText("10")).toBeInTheDocument();
    expect(screen.getByText("3")).toBeInTheDocument();
    expect(screen.getByText("1")).toBeInTheDocument();

    // Default labels
    expect(screen.getByText(/Up to date/i)).toBeInTheDocument();
    expect(screen.getByText(/Updates available/i)).toBeInTheDocument();
    expect(screen.getByText(/Failed/i)).toBeInTheDocument();
  });

  it("matches snapshot layout", () => {
    const { container } = render(
      <InstanceStatusOverview upToDate={7} updatesAvailable={2} failed={0} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

