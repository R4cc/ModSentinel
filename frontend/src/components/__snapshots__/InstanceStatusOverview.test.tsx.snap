// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InstanceStatusOverview > matches snapshot layout 1`] = `
<div
  aria-label="Instance status overview"
  class="grid grid-cols-1 gap-sm sm:grid-cols-3 max-w-3xl sm:max-w-4xl"
  role="region"
>
  <div
    class="rounded-md border p-sm shadow-sm"
  >
    <div
      class="flex items-center gap-sm"
    >
      <span
        aria-hidden="true"
        class="shrink-0 text-emerald-600"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle-check-big h-6 w-6"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.801 10A10 10 0 1 1 17 3.335"
          />
          <path
            d="m9 11 3 3L22 4"
          />
        </svg>
      </span>
      <div
        class="flex items-baseline gap-xs"
      >
        <span
          class="text-2xl font-bold leading-none text-emerald-700"
        >
          7
        </span>
        <span
          class="text-sm text-muted-foreground"
        >
          Up to date
        </span>
      </div>
    </div>
  </div>
  <div
    class="rounded-md border p-sm shadow-sm"
  >
    <div
      class="flex items-center gap-sm"
    >
      <span
        aria-hidden="true"
        class="shrink-0 text-amber-600"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-refresh-ccw h-6 w-6"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 12a9 9 0 0 0-9-9 9.75 9.75 0 0 0-6.74 2.74L3 8"
          />
          <path
            d="M3 3v5h5"
          />
          <path
            d="M3 12a9 9 0 0 0 9 9 9.75 9.75 0 0 0 6.74-2.74L21 16"
          />
          <path
            d="M16 16h5v5"
          />
        </svg>
      </span>
      <div
        class="flex items-baseline gap-xs"
      >
        <span
          class="text-2xl font-bold leading-none text-amber-700"
        >
          2
        </span>
        <span
          class="text-sm text-muted-foreground"
        >
          Updates available
        </span>
      </div>
    </div>
  </div>
  <div
    class="rounded-md border p-sm shadow-sm"
  >
    <div
      class="flex items-center gap-sm"
    >
      <span
        aria-hidden="true"
        class="shrink-0 text-red-600"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle-x h-6 w-6"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="m15 9-6 6"
          />
          <path
            d="m9 9 6 6"
          />
        </svg>
      </span>
      <div
        class="flex items-baseline gap-xs"
      >
        <span
          class="text-2xl font-bold leading-none text-red-700"
        >
          0
        </span>
        <span
          class="text-sm text-muted-foreground"
        >
          Failed
        </span>
      </div>
    </div>
  </div>
</div>
`;
